src/main/java/com/google/devtools/build/lib/bazel/repository/PatchUtil.java

```diff
@@ -417,10 +417,25 @@ private static void checkFilesStatusForPatching(
    *
    * @param patchFile the patch file to apply
    * @param strip the number of leading components to strip from file path in the patch file
-   * @param outputDirectory the repository directory to apply the patch file
+   * @param outputDirectory the directory to apply the patch file to
    */
   public static void apply(Path patchFile, int strip, Path outputDirectory)
       throws IOException, PatchFailedException {
+    apply(patchFile, strip, outputDirectory, /* singleFile= */ null);
+  }
+
+  /**
+   * Apply a patch file under a directory
+   *
+   * @param patchFile the patch file to apply
+   * @param strip the number of leading components to strip from file path in the patch file
+   * @param outputDirectory the directory to apply the patch file to
+   * @param singleFile if not null, only apply the parts of the patch file that apply to this file.
+   *     Renaming the file is not supported in this case.
+   */
+  public static void apply(
+      Path patchFile, int strip, Path outputDirectory, @Nullable Path singleFile)
```

Comment:
this new optional parameter changes the method's behavior quite significantly. Could we rename this method to `applyToSingleFile` and make the `@Nullable` version private? (and maybe rename `singleFile` the parameter to `fileToPatch` or something)